File: golang/architecture-shell-cqrs/tests/query_test.go
package tests

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/universalddd/architecture-shell-cqrs"
)

// T039: Should_ReturnCorrectType_When_QueryHandlerExecutes
func TestShould_ReturnCorrectType_When_QueryHandlerExecutes(t *testing.T) {
	// Given: A query and its handler
	query := &GetTestDataQuery{ID: "test-123"}
	handler := &GetTestDataQueryHandler{}

	handlers := map[string]interface{}{
		"GetTestDataQuery": handler,
	}
	behaviors := []interface{}{}

	mediator := cqrs.NewMediator(handlers, behaviors)
	ctx := context.Background()

	// When: Query is executed through mediator
	result, err := mediator.Send(ctx, query)

	// Then: Should return correct result type
	assert.NoError(t, err)
	assert.NotNil(t, result)

	// Verify the result is of expected type
	dto, ok := result.(*TestDataDto)
	assert.True(t, ok, "Result should be of type *TestDataDto")
	assert.Equal(t, "test-123", dto.ID)
	assert.Equal(t, "Test Data", dto.Name)
}

// Test helper types for query tests
type GetTestDataQuery struct {
	ID string
}

func (q *GetTestDataQuery) IsRequest() {}
func (q *GetTestDataQuery) IsQuery()   {}

type TestDataDto struct {
	ID   string
	Name string
}

type GetTestDataQueryHandler struct{}

func (h *GetTestDataQueryHandler) Handle(ctx context.Context, query *GetTestDataQuery) (*TestDataDto, error) {
	return &TestDataDto{
		ID:   query.ID,
		Name: "Test Data",
	}, nil
}

File: typescript-nodejs/architecture-shell-cqrs/tests/unit/query.test.ts
import 'reflect-metadata';
import { Mediator } from '../../src/Mediator';
import { Query } from '../../src/Query';
import { IQueryHandler } from '../../src/IQueryHandler';

/**
 * UT-002: Query Return Type Contracts
 *
 * These tests verify that query handlers return correct types.
 * Following TDD principles: tests MUST fail before implementation.
 */

// Test DTO
interface OrderDetailsDto {
  orderId: string;
  customerName: string;
  totalAmount: number;
}

// Test query
class GetOrderDetailsQuery implements Query<OrderDetailsDto> {
  _isQuery = true as const;
  __phantom?: OrderDetailsDto;
  constructor(public readonly orderId: string) {}
}

describe('QueryTests', () => {
  // UT-002: T040
  test('Should_ReturnCorrectType_When_QueryHandlerExecutes', async () => {
    // Given: A query handler that returns a specific DTO type
    const expectedDto: OrderDetailsDto = {
      orderId: '123',
      customerName: 'John Doe',
      totalAmount: 100.50,
    };

    const handler: IQueryHandler<GetOrderDetailsQuery, OrderDetailsDto> = {
      async handle(_request: GetOrderDetailsQuery, _signal: AbortSignal): Promise<OrderDetailsDto> {
        return expectedDto;
      },
    };

    const mediator = new Mediator(
      [{ requestType: GetOrderDetailsQuery, handler }]
    );

    // When: Query is executed through mediator
    const signal = new AbortController().signal;
    const result = await mediator.send(new GetOrderDetailsQuery('123'), signal);

    // Then: Should return the correct TResult type
    expect(result).toEqual(expectedDto);
    expect(result.orderId).toBe('123');
    expect(result.customerName).toBe('John Doe');
    expect(result.totalAmount).toBe(100.50);
  });
});
